### PR TITLE
SMB2 Rename and Delete Improvement

### LINKED ISSUE #113: Improve SMB2 rename and delete logic
SMB2 rename and delete functionalities, implemented in the File class, are both using `SetInfoResponse` packet with `FileRenameInformation` and `FileDispositionInformation` FSCC File Information classes.
However, the File Information classes are instantiated in the File methods and used to set the packet's attribute:
- https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/smb2/file.rb#L223
- https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/smb2/file.rb#L149

This design can be improved by simply using the File Information code after instantiating `SetInfoResponse` and let the this class set the related attributes. As a reference, the SMB1 TRANS2 implementation is currently using this design to implement sub-commands (https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/smb1/packet/trans2/request.rb and https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/smb1/packet/trans2/subcommands.rb).

### PR DESCRIPTION
**Pull request by chrisdlf**
This PR updates SMB2 rename and delete functionalities.
I also needed to update SMB1 and SMB2 Tree list methods.

This fixes #113.

# Verification Steps
- [x] `bundle install`
- [x] Verify you can rename a file with `ruby examples/rename_file.rb <ip-address> <username> <password> <share> <filename> <new filename>`
- [x] Verify you can delete a file with `ruby examples/delete_file.rb <ip-address> <username> <password> <share> <filename>`
- [x] Verify you can list files from a directory with `ruby examples/list_directory.rb <ip-address> <username> <password> <share> <directory>`
## `rake spec`
- [x] `rake spec`
- [x] VERIFY no failures

### CODE REVIEW

File: lib/ruby_smb/smb2/info_type.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    # [2.2.37 SMB2 QUERY_INFO Request](https://msdn.microsoft.com/en-us/library/cc246557.aspx)
+    module InfoType
+      # The file information
+      SMB2_0_INFO_FILE       = 0x01
+
+      # The underlying object store information
+      SMB2_0_INFO_FILESYSTEM = 0x02
+
+      # The security information
+      SMB2_0_INFO_SECURITY   = 0x01
```
Comment (dmohanty-r7) [inline lib/ruby_smb/smb2/info_type.rb]: This should be 0x03

File: lib/ruby_smb/smb2/info_type.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+      # The underlying object store quota information
+      SMB2_0_INFO_QUOTA      = 0x01
```
Comment (dmohanty-r7) [inline lib/ruby_smb/smb2/info_type.rb]: This should be 0x04

File: lib/ruby_smb/smb2/info_type.rb (hunk 1 above)
Comment (chrisdlf) [reply lib/ruby_smb/smb2/info_type.rb]: Good catch! A bad copy/paste, thanks!

Comment (chrisdlf) [reply lib/ruby_smb/smb2/info_type.rb]: Addressed in dcd7ff20e8253df592e5358d503d916d62fb479a

File: lib/ruby_smb/field/smb_gea_list.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ module Field
     # [2.2.1.2.1.1 SMB_GEA_LIST](https://msdn.microsoft.com/en-us/library/ff359447.aspx)
     class SmbGeaList < BinData::Record
       endian :little
-      uint32  :size_of_list, label: 'Size of List in Bytes', initial_value: -> { gea_list.do_num_bytes }
+      uint32  :size_of_list, label: 'Size of List in Bytes', initial_value: -> { self.do_num_bytes }
```
Comment (chrisdlf) [inline lib/ruby_smb/field/smb_gea_list.rb]: This change was needed to follow the protocol: https://msdn.microsoft.com/en-us/library/ff359447.aspx

File: lib/ruby_smb/smb1/packet/trans2/find_first2_request.rb [hunk 4]
```diff
@@ -38,7 +42,8 @@ def length
 
           # The Trans2 Data Blcok for this particular Subcommand
           class Trans2Data < BinData::Record
-            smb_gea_list :extended_attribute_list, label: 'Get Extended Attribute List'
+            smb_gea_list :extended_attribute_list, label: 'Get Extended Attribute List',
+              onlyif: -> { parent.trans2_parameters.information_level == FindInformationLevel::SMB_INFO_QUERY_EAS_FROM_LIST}
```
Comment (chrisdlf) [inline lib/ruby_smb/smb1/packet/trans2/find_first2_request.rb]: Please see the Trans2_Data section of https://msdn.microsoft.com/en-us/library/ee441987.aspx

**Pull request merged**